src/ott/solvers/linear/sinkhorn.py

```diff
@@ -797,7 +800,10 @@ def body_fn(
       state: SinkhornState, compute_error: bool
   ) -> SinkhornState:
     ot_prob, solver = const
-    return solver.one_iteration(ot_prob, state, iteration, compute_error)
+    state = solver.one_iteration(ot_prob, state, iteration, compute_error)
+    if solver.progress_fn is not None:
+        host_callback.id_tap(solver.progress_fn, (iteration, solver.max_iterations, state.errors))
```

Comment:
I think you need to pass `solver.inner_iterations` to the `progress_fn`, and then the error index you want to print is not `[-1]` (`errors` is a static vector of size `max_iterations`) but instead `[iteration//inner_iterations]` (to be more efficient the error is only recomputed every `inner_iterations`).

Reply:
Cool! Indeed, I remember your advice, but I did not know how to code it and did not call your attention onto it again. Great catch, thanks!

This will be taken into account in my next commit

Reply:
I think it might be beneficial to pass the whole state to the callback, but am afraid `host_callback` only allows primitives as `jax` arrays atm.

Reply:
Just bumped into this... Have you considered pure_callback or debug.callback? I believe the host_callback suite is largely deprecated.

Reply:
Thank you both.

@michalk8 Indeed I find it awkward to pass that many parameters.

@ntenenz No, I was not aware of these mechanisms, thanks, I'll look into it!

Reply:
Thank you both!

@michalk8 Yes I find it awkward to pass that many parameters. Hopefully what @ntenenz mentions could bring a solution. I'll look into it!

Reply:
@ntenenz I think that both [`jax.pure_callback`](https://jax.readthedocs.io/en/latest/_autosummary/jax.pure_callback.html) and [`jax.debug.callback`](https://jax.readthedocs.io/en/latest/_autosummary/jax.debug.callback.html) cannot be used for functions which have side-effects like printing or updating a progress bar, making them not really fit here. Did I miss something?

Reply:
I think it depends on what you mean by stateful here and how you expect the callback to perform under function transforms (e.g., `vmap`, `grad`, etc.). Perhaps a few questions may help illustrate what I'm talking about:
* Would it be ok if the callback is invoked multiple times with the same arguments (e.g., with a progress bar, you'd simply be setting the progress to the same value so it may be fine)?
* Would it be ok if the callback's invocation is reordered so that it's not immediately invoked upon iteration completion (i.e., timing values may not be precise)?

From the [jax docs](https://jax.readthedocs.io/en/latest/debugging/print_breakpoint.html#more-control-with-jax-debug-callback), it indicates that `jax.debug.print` is actually just a thin convenience wrapper around `jax.debug.callback` so printing should hopefully (?) be fine.

Reply:
Thanks again @ntenenz, the idea being essentially to report a rough estimate of the progress, printing is sufficient indeed.

@marcocuturi @michalk8 @ntenenz would the following be sufficient?

<img width="539" alt="image" src="https://user-images.githubusercontent.com/1936828/211042653-4d51b1bd-4a75-43a9-98fb-0f61c506f1fd.png">

Reply:
The user does not have to provide a callback function. The benefit is that the report can be adjusted depending on the use case without user intervention.